Raft test topology part 2

alecco opened the pull request:
Give cluster control to pytests. While there add missing stop gracefully and add server to ScyllaCluster.
Clusters can be marked dirty but they are not recycled yet. This will be done in a later series.

- `ClusterManager` is now `@asynccontextmanager`   (Kamil/Kostja)
- `Cluster` is handled inside `ClusterManager` (and it will change)
- add docstring to `add_server()` in the proper commit
- remove `take_log_savepoint()` added in `before_test()` (Kamil)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1686/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1693/

kbr- commented on test/topology/conftest.py:
```diff
@@ -48,6 +52,44 @@ def run_async(self, *args, **kwargs) -> asyncio.Future:
+# --port options, and defaults to localhost and 9042, respectively.
+# We use scope="session" so that all tests will reuse the same client object.
+@pytest.fixture(scope="session")
+def cql(request):
```
Using the existing `cql` fixture from test/conftest.py doesn't work? Wouldn't the server just ignore the auth data with AllowAll?

nyh replied:
Like Cato the Elder, I want to state again my belief that the top-level test/conftest.py should not exist. There can be common libraries that many test suites reuse, but there will always be differences. I think this is becoming more obvious now that you have a "cql" fixture in both test/conftest.py and test/topology/conftest.py and nominally both apply and you need to start guessing which one is relevant.

kbr- replied:
OK, if the end-game is that we remove the top-level one, then it makes sense.

kbr- commented on test/pylib/scylla_cluster.py:
```diff
@@ -485,3 +487,41 @@ def after_test(self, name) -> None:
+    manager = ScyllaClusterManager(test_name, clusters)
+    await manager.start()
+    yield manager
+    await manager.stop()
```
Don't we need to wrap this in `try...finally`? The examples in documentation do this:
https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager

kbr- commented on test.py:
```diff
@@ -809,6 +827,41 @@ def write_junit_failure_report(self, xml_res: ET.Element) -> None:
+    async def run(self, options: argparse.Namespace) -> Test:
+
+        async with get_cluster_manager(self.shortname, self.suite.clusters) as manager:
+            self.args.insert(0, "--host={}".format(manager.cluster[0].host))
```
Nit: a local `cluster = manager.cluster` variable would make the code more readable.

alecco replied:
The cluster can change inside manager. For topology tests this will be more often than not. The exception handling code later could have a dangling cluster reference and would bring irrelevant logs.

kbr- replied:
I know that the cluster object can mutate. But can `manager.cluster` really change, i.e. point to one cluster first, then another cluster - during the same test? How could this happen? And if this can happen - won't we have some big problems? Because that would bean we could call `before_test` on a different object than `after_test` and so on.

kbr- commented:
1. the PR cover letter should be mostly self-containing, it should explain what the PR does. It will be part of the merge commit message. I think what you had in the original PR fits here:
>Give cluster control to pytests. While there add missing stop gracefully and add server to ScyllaCluster.
> Clusters can be marked dirty but they are not recycled yet. This will be done in a later series.

2. I would love a summary of the changes (better than "minor fixes") or a diff at least.

kbr- commented:
Seems many of the comments from my previous review weren't addressed. Please address them either by applying a change, or explaining that you won't change it and why you decided not to change it.

- comments for some fields in ScyllaCluster (https://github.com/scylladb/scylladb/pull/10815#discussion_r935629490)
- removing self.cluster (https://github.com/scylladb/scylladb/pull/10815#discussion_r935634302)
- inefficient `__getitem__` - if there is no better way (e.g. removing `__getitem__` altogether), tell me. (https://github.com/scylladb/scylladb/pull/10815#discussion_r935635653)
- comment `for server in self.running.values()` (https://github.com/scylladb/scylladb/pull/10815#discussion_r935637668)
- calculation of `last_seed` (https://github.com/scylladb/scylladb/pull/10815#discussion_r935642671)
- named tuples for return values of server_stop/start etc. (https://github.com/scylladb/scylladb/pull/10815#discussion_r935647178)
- implementing server_restart in terms of server_stop+server_start - can it be done? (https://github.com/scylladb/scylladb/pull/10815#discussion_r935649287)

alecco commented:
v2 changes (all @kbr-)

- Added comments to `ScyllaCluster` fields `.running` `.stopped` `.removed` `.is_running`
- Removed leftover `self.cluster`
- Added FIXME in `test.py` for calls of `ScyllaCluster.__getitem()` as the logic is not sound, and we can remove it altogether
- Fixed seed logic, now all running servers are passed; this avoids tracking `last_seed`; empty list instead of `None`
- Use a `NamedTuple` as server action return value (e.g. `ScyllaCluster.stop_server()`)
- Refactor `ScyllaCluster.restart_server()` to use `ScyllaCluster.stop_server()` and `ScyllaCluster.start_server()`
- Fix `@asynccontextmanager` `get_cluster_manager()` to use `try: finally:`

nyh commented on test.py:
```diff
@@ -409,6 +409,24 @@ def pattern(self) -> str:
         return "*test.cql"
 
 
+class TopologyTestSuite(PythonTestSuite):
```
We have a topology/ subdirectory, why can't the extra logic for topology tests be there, and not in the main test.py?
I think we're putting too much code in test.py, and there is a lot of duplication in test.py and it makes it harder to fix things or to understand which code controls what.
But I won't insist, I understand that we have different philosophies regarding test.py.

nyh replied:
Moreover, the comment below doesn't help me understand how "tests dealing with topology changes" are different from other suite types and why we need a separate test suite type for it. Is the difference that these tests run on multi-node clusters and not one node? Is the difference that these tests each *changes* the cluster, so we need a new cluster for each test? Or something else?

nyh commented on test/pylib/scylla_cluster.py:
```diff
@@ -488,6 +489,7 @@ def _seeds(self) -> List[str]:
     async def add_server(self) -> str:
         """Add a new server to the cluster"""
         server = self.create_server(self.name, self._seeds())
+        self.is_dirty = True
```
I have some doubts about the concept of "dirty". What makes a cluster "dirty"? Is a cluster which started with 3 nodes, one was added and then one was removed, "dirty"? Why?

kbr- replied:
I also requested a comment explaining what `is_dirty` means, among some other fields, I see some of the ones I requested were commented upon but not all, including `is_dirty`:
https://github.com/scylladb/scylladb/pull/10815#discussion_r935629490

> Is a cluster which started with 3 nodes, one was added and then one was removed, "dirty"? Why?

I guess any topology change makes a cluster dirty. We definitely want to mark cluster dirty if it's missing a node, for example. In a follow-up we may think about optimizations (but let's not do this right now), e.g. maybe if the number of nodes is the same before and after we can leave it "not dirty" (but I'm not sure).

kbr- replied:
Actually, `is_dirty` is commented upon now:
```
        # cluster was modified in a way it should not be used in subsequent tests
        self.is_dirty: bool = False
```

The discussion what modifications should make a cluster dirty... We could expand on it in the comments, but meh, I don't think it matters that much.

We can improve on this later.

nyh replied:
The reason I asked for a comment is that it's not clear what "cluster was modified" really means. Here is a concrete example that led me to this question: A lot of existing dtests create a three-node cluster, then kill one node, and do something, and then bring back the third node to life and test what happened. So when the test has ended, we again have a working three-node cluster. Is such a cluster "dirty" or not "dirty"? Should it be used in subsequent tests, or not?

But I can accept that for now, we don't know exactly what "is_dirty" will be - and that it will continue to evolve in the near future. If this is the case, then indeed the comment we have is good enough.

kbr- reviewed (approved):
LGTM, I'll merge when CI passes.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -364,6 +364,27 @@ async def stop(self) -> None:
+        except ProcessLookupError:
+            pass
+        else:
+            await self.cmd.wait()
```
I'll leave a FIXME here when merging.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1740/

kbr- commented:
Failure unrelated to PR, restarted CI

alecco commented:
@nyh 
> Like Cato the Elder, I want to state again my belief that the top-level test/conftest.py should not exist. There can be common libraries that many test suites reuse, but there will always be differences. I think this is becoming more obvious now that you have a "cql" fixture in both test/conftest.py and test/topology/conftest.py and nominally both apply and you need to start guessing which one is relevant.

I moved things out of it and into `test/topology/conftest.py` already in part 1. As I stated in an earlier comment, I moved the `cql` fixture to make it easier to see the subsequent changes to the code (in the same file). I plan to get rid of it altogether later on (for topology tests). But I want to leave a trail of small changes easy to follow.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1746/

kbr- commented:
Qd
